### PR TITLE
Added: Color box method

### PR DESCRIPTION
**Pull request by dvdvideo1234**

### CODE REVIEW

**Review by robotboy655 — CHANGES_REQUESTED** [review]
Since the panel the function creates is Sandbox exclusive, this function should be added to a Sandbox exclusive panel as well.

See this commit:
https://github.com/Facepunch/garrysmod/commit/b1816ba491a11b4155c7e23e729c4a19a59b0ee7

**Comment by dvdvideo1234** [comment]
Added those two to DForm also.

File: garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua [hunk 1]
```diff
@@ -91,6 +91,27 @@ function PANEL:KeyBinder( label1, convar1, label2, convar2 )
 
 end
 
+function PANEL:ColorBox(label, convarr, convarg, convarb, convara)
```
Comment (robotboy655) [inline garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua:94]: Coding style: Add spaces: `( label, ... , convara )`

Also possibly rename to `ColorPicker`?

File: garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua [hunk 2]
```diff
@@ -91,6 +91,27 @@ function PANEL:KeyBinder( label1, convar1, label2, convar2 )
 
 end
 
+function PANEL:ColorBox(label, convarr, convarg, convarb, convara)
+
+	local color = vgui.Create("CtrlColor", self)
```
Comment (robotboy655) [inline garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua:96]: Coding style

File: garrysmod/lua/vgui/dform.lua [hunk 3]
```diff
@@ -134,6 +134,61 @@ function PANEL:PropSelect( label, convar, models, height )
 
 end
 
+function PANEL:ToolPresets( group, cvarlist )
```
Comment (robotboy655) [inline garrysmod/lua/vgui/dform.lua:137]: Get rid of all changes in this file. We cannot put Sandbox exclusive code to DForm.

**Review by robotboy655 — CHANGES_REQUESTED** [review]

File: garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua [hunk 4]
```diff
@@ -72,9 +57,61 @@ function PANEL:MatSelect( strConVar, tblOptions, bAutoStretch, iWidth, iHeight )
...
---[[---------------------------------------------------------
-	Name: FillViaTable
------------------------------------------------------------]]
+function PANEL:ToolPresets( group, cvarlist )
+
+	local preset = vgui.Create( "ControlPresets", self )
+
+	preset:SetPreset( group )
+	preset:AddOption( "#preset.default", cvarlist )
+	for key, val in pairs( table.GetKeys( cvarlist ) ) do
```
Comment (GitSparTV) [inline garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua]: Use `ipairs`. table.GetKeys returns an array

File: garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua (hunk 1 above)
Comment (dvdvideo1234) [reply garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua:94]: Sure thing

File: garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua (hunk 2 above)
Comment (dvdvideo1234) [reply garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua:96]: Done

File: garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua (hunk 4 above)
Comment (dvdvideo1234) [reply garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua]: Done

File: garrysmod/lua/vgui/dform.lua (hunk 3 above)
Comment (dvdvideo1234) [reply garrysmod/lua/vgui/dform.lua]: Done in https://github.com/Facepunch/garrysmod/pull/1854/commits/b964f8796ea9c5184a31894da9ff2fef6b2a66f6

File: garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua [hunk 5]
```diff
@@ -91,6 +91,27 @@ function PANEL:KeyBinder( label1, convar1, label2, convar2 )
...
 end
 
+function PANEL:ColorPicker( label, convarr, convarg, convarb, convara )
+
+	local color = vgui.Create( "CtrlColor", self )
+
+	color:Dock( TOP )
+	color:SetLabel( label )
+
+	color:SetConVarR( convar )
```
Comment (robotboy655) [inline garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua:101]: BUG: `convar` doesn't exist. Probably should name these something like `convarR`, `convarG`, etc so its easier to see.

File: garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua [hunk 6]
```diff
@@ -63,7 +63,7 @@ function PANEL:ToolPresets( group, cvarlist )
 
 	preset:SetPreset( group )
 	preset:AddOption( "#preset.default", cvarlist )
-	for key, val in pairs( table.GetKeys( cvarlist ) ) do
+	for key, val in ipairs( table.GetKeys( cvarlist ) ) do
```
Comment (robotboy655) [inline garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua:66]: Why does this even use `table.GetKeys`?

Wouldn't this be better?

```lua
for cvar, defVal in pairs( cvarlist ) do
preset:AddConVar( cvar )
end
```

**Review by robotboy655 — CHANGES_REQUESTED** [review]

File: garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua (hunk 6 above)
Comment (dvdvideo1234) [reply garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua:66]: The idea is to use https://wiki.facepunch.com/gmod/Tool:BuildConVarList but I guess there is no point if they are ordered or not.

Comment (GitSparTV) [reply garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua:66]: GetKeys doesn't make them ordered either

Comment (dvdvideo1234) [reply garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua]: Yep, whatever comes first becomes 1... :D Fixed in https://github.com/Facepunch/garrysmod/pull/1854/commits/e57c7cbf70399d62161f5e185aabc2397ee693d2

File: garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua (hunk 5 above)
Comment (dvdvideo1234) [reply garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua]: Fixed: https://github.com/Facepunch/garrysmod/pull/1854/commits/e57c7cbf70399d62161f5e185aabc2397ee693d2

File: garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua (hunk 6 above)
Comment (dvdvideo1234) [reply garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua]: https://github.com/Facepunch/garrysmod/blob/master/garrysmod/gamemodes/sandbox/entities/weapons/gmod_tool/stool.lua#L67

File: garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua [hunk 7]
```diff
@@ -91,6 +91,27 @@ function PANEL:KeyBinder( label1, convar1, label2, convar2 )
 
 end
 
+function PANEL:ColorPicker( label, convar_r, convar_g, convar_b, convar_a )
```
Comment (GitSparTV) [inline garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua:94]: I think the code style uses camelCase, not underscored_case

**Review by GitSparTV — APPROVED** [review]

**Comment by dvdvideo1234** [comment]
This is perfect @GitSparTV @robotboy655 ;) Ready to be merged 👍

**Comment by GitSparTV** [comment]
Rubat, should the changes outside of PR scope be in the name?
Like, this PR adds new function, but also changes other things little bit.

**Comment by GitSparTV** [comment]
I didn't say it's wrong. I think just mentioning this in the PR name is fine. Just wait for Rubat to answer

**Comment by GitSparTV** [comment]
Is there a semantic difference? How would you put false in it

**Comment by dvdvideo1234** [comment]
It will just error when the user passes false instead of convar list. Looked sane to me at the time :D

File: garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua [hunk 8]
```diff
@@ -46,9 +46,8 @@ function PANEL:MatSelect( strConVar, tblOptions, bAutoStretch, iWidth, iHeight )
 
 	if ( tblOptions != nil ) then
 		for k, v in pairs( tblOptions ) do
-			local label = k
-			if ( isnumber( label ) ) then label = v end
-			MatSelect:AddMaterial( label, v )
+			local nam = isnumber( k ) and v or k
+			MatSelect:AddMaterial( nam, v )
```
Comment (robotboy655) [inline garrysmod/gamemodes/sandbox/gamemode/spawnmenu/controlpanel.lua:50]: You really need to learn to stop including unrelated commits to a Pull Request. I will let it slide this time.

**Review by robotboy655 — APPROVED** [review]

**Pull request merged**